src/context/handle_manager.rs

```diff
@@ -0,0 +1,133 @@
+        }
+    }
+
+    /// Adds a handle to the HandleManager
+    pub(super) fn add_handle(
+        &mut self,
+        handle: ObjectHandle,
+        handle_drop_action: HandleDropAction,
+    ) -> Result<()> {
+        if handle == ObjectHandle::None || handle == ObjectHandle::Null {
+            error!("Handle manager does not handle None or Null handles");
+            return Err(Error::local_error(WrapperErrorKind::InvalidParam));
+        }
+
+        if self.open_handles.contains_key(&handle) {
+            error!("Handle({}) is already open", ESYS_TR::from(handle));
+            return Err(Error::local_error(WrapperErrorKind::InvalidHandleState));
+        }
+        let _ = self.open_handles.insert(handle, handle_drop_action);
+        Ok(())
+    }
+
+    /// Sets the handle as flushed which removes it from the manager.
+    ///
+    /// # Error
+    /// If the handle was not set to be flushed then this will cause an
+    /// error but the handle will still be removed from the handler.
+    pub(super) fn set_as_flushed(&mut self, handle: ObjectHandle) -> Result<()> {
+        self.open_handles
+            .remove(&handle)
+            .ok_or_else(|| {
+                error!("Handle({}) does not exist", ESYS_TR::from(handle));
+                Error::local_error(WrapperErrorKind::InvalidHandleState)
+            })
+            .and_then(|handle_drop_action| {
+                if handle_drop_action == HandleDropAction::Flush {
+                    Ok(())
+                } else {
+                    error!(
+                        "Flushing handle({}) that should not have been flushed.",
+                        ESYS_TR::from(handle)
+                    );
+                    Err(Error::local_error(WrapperErrorKind::InvalidHandleState))
+                }
+            })
+    }
+
+    /// Sets the handles as closed which removes it from the handler.
+    ///
+    /// # Error
+    /// If the handle was set to be flushed then this will cause an
+    /// error but the handle will still be removed from the handler.
+    pub(super) fn set_as_closed(&mut self, handle: ObjectHandle) -> Result<()> {
+        self.open_handles
+            .remove(&handle)
+            .ok_or_else(|| {
+                error!("Handle({}) does not exist", ESYS_TR::from(handle));
+                Error::local_error(WrapperErrorKind::InvalidHandleState)
+            })
+            .and_then(|handle_drop_action| {
+                if handle_drop_action == HandleDropAction::Close {
+                    Ok(())
+                } else {
+                    error!(
+                        "Closing handle({}) that should have been flushed",
+                        ESYS_TR::from(handle)
+                    );
+                    Err(Error::local_error(WrapperErrorKind::InvalidHandleState))
+                }
+            })
+    }
+
+    /// Retrieves all handles that needs to be flushed.
+    pub(super) fn handles_to_flush(&self) -> Vec<ObjectHandle> {
```

Comment:
I'm wondering whether we really need to build a new vector. Given that the only use for this is to iterate over it, maybe instead just return an `Iter<ObjectHandle>`?

Reply:
It was just the easiest way to not hold a reference to the manager while calling close/flush because those functions also interacts with the manager. I will try to improve on it.

Reply:
Eh, I don't care that much (it's only upon drop, and not that invasive), just figured it might be nicer.